src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs

```diff
@@ -848,10 +849,246 @@ public void ReadingCorruptedStreamThrows()
+        {
+            // BuildErrorEventArgs error = new("Subcategory", "Code", "File", 1, 2, 3, 4, "Message", "HelpKeyword", "SenderName");
+            BuildErrorEventArgs error = new(null, null, null, 1, 2, 3, 4, null, null, null);
+            BuildFinishedEventArgs finished = new("Message", "HelpKeyword", true);
+
+            var memoryStream = new MemoryStream();
+            var binaryWriter = new BinaryWriter(memoryStream);
+            var binaryReader = new BinaryReader(memoryStream);
+            var buildEventArgsWriter = new BuildEventArgsWriter(binaryWriter);
+
+            buildEventArgsWriter.Write(error);
+
+            int positionAfterFirstEvent = (int)memoryStream.Position;
+            memoryStream.Position = 0;
+            // event type
+            Microsoft.Build.Shared.BinaryReaderExtensions.Read7BitEncodedInt(binaryReader);
+            int eventSizePos = (int)memoryStream.Position;
+            int eventSize = Microsoft.Build.Shared.BinaryReaderExtensions.Read7BitEncodedInt(binaryReader);
+            int positionAfterFirstEventSize = (int)memoryStream.Position;
+            memoryStream.Position = eventSizePos;
+            // simulate there are 4 bytes less in the future version of the event - while our reader expects those
+            Microsoft.Build.Shared.BinaryWriterExtensions.Write7BitEncodedInt(binaryWriter, eventSize - 4);
+            memoryStream.Position.ShouldBe(positionAfterFirstEventSize, "The event size need to be overwritten in place - without overwriting any bytes after the size info");
+            // remove the 4 bytes - so that actual size of event is inline with it's written size.
+            memoryStream.Position = positionAfterFirstEvent - 4;
+
+            buildEventArgsWriter.Write(finished);
+
+            // Remember num of bytes written - we should read them all.
+            long length = memoryStream.Length;
+            // Now move back to the beginning of the stream and start reading.
+            memoryStream.Position = 0;
+
+            using var buildEventArgsReader = new BuildEventArgsReader(binaryReader, BinaryLogger.FileFormatVersion)
+            {
+                SkipUnknownEvents = true
+            };
+
+            List<(ReaderErrorType errorType, BinaryLogRecordKind recordKind, string error)> readerErrors = new();
+            buildEventArgsReader.OnRecoverableReadError += (t, et, e) => readerErrors.Add((t, et, e()));
+
+            var deserializedEvent = buildEventArgsReader.Read();
+
+            readerErrors.Count.ShouldBe(1);
+            readerErrors[0].errorType.ShouldBe(ReaderErrorType.UnknownFormatOfEventData);
+            readerErrors[0].recordKind.ShouldBe(BinaryLogRecordKind.Error);
+            // Expect StreamChunkOverReadException or EndOfStreamException - based on the implementation
+            //  of indicating attempt to overread in the TransparentReadStream.
+            readerErrors[0].error.ShouldContain("StreamChunkOverReadException");
```

Comment:
This seems a bit user-unfriendly, is there a way we could offer a cleaner exception?

Reply:
@rainersigwald - Can you please elaborate? I do not understand what you meant ... which possibly be given by a bit confusing comment I put here :-/

But in any case - the error currently looks as such:

```
BuildEvent record number 0 (serialized size: 15) attempted to perform disallowed reads (details: Microsoft.Build.Logging.StreamChunkOverReadException: Attempt to read 1 bytes, when only 0 are allowed to be read.). Skipping the record.
```

However I'm now leaning towards getting rid of the `StreamChunkOverReadException` alltogether - as the extra info it brings is useless to user and it can be seen as contracts breaking (see discussion for the details: https://github.com/dotnet/msbuild/pull/9307#discussion_r1362026217).
The error in version without `StreamChunkOverReadException` would look as such:

```
BuildEvent record number 0 (serialized size: 15) attempted to perform disallowed reads (details: System.IO.EndOfStreamException: Unable to read beyond the end of the stream.). Skipping the record.
```

Would you prefer more/less/different information in the error?

Reply:
In the end I stuck to `EndOfStreamException` offering the ability for little cleaner implementation. The error that will be provided to user in case of over-read attempt will be:

```
BuildEvent record number 0 (serialized size: 15) attempted to perform disallowed reads (details: System.IO.EndOfStreamException: Unable to read beyond the end of the stream.). Skipping the record.
```

Please unresolve this discussion if you feel it is not user-friandly.